[Xcode] Move header migration logic to a script phase

emw-apple opened the pull request:
#### 36cefe446ec77fb3dd2e3949f083868b6679c54f
<pre>
[Xcode] Move header migration logic to a script phase
<a href="https://bugs.webkit.org/show_bug.cgi?id=251635">https://bugs.webkit.org/show_bug.cgi?id=251635</a>

Reviewed by Alexey Proskuryakov.

WebKit and WebKitLegacy &quot;migrate&quot; headers from other frameworks which
they reexport symbols from. On Mac, some of these migrated headers are
public (namely the WebKitLegacy API on that platform), but on all
embedded platforms the migrated headers are private. Xcode doesn&apos;t
natively support headers that have different visibility per-platform.

Previously, we cheated by using a build setting to &quot;redirect&quot; migrated
headers into the PrivateHeaders/ directory on embedded platforms.
Unfortunately, this breaks tapi&apos;s SRCROOT mode, so we need to implement
a different workaround.

Replace the migrated header references in both projects with a script
phase that calls a new script, migrate-headers.py. This script reads
from a filelist to process headers and write them to the appropriate
headers folder. Headers that are written to
$(WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR) are the ones with per-platform
visibility.

* Configurations/WebKitProjectPaths.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/MigratedHeaders-input.xcfilelist: Added.
* Source/WebKit/MigratedHeaders-output.xcfilelist: Added.
* Source/WebKit/Scripts/migrate-headers.py: Added.
* Source/WebKit/Scripts/postprocess-header-rule: Now, migrated headers
  that this script sees come from
  $(BUILT_PRODUCTS_DIR)/WebKitLegacy.framework, instead of
  WebKitLegacy&apos;s SRCROOT. This means they have already been
  postprocessed by WebKitLegacy. Remove the logic that is duplicated by
  WebKitLegacy&apos;s postprocessing script.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259916@main">https://commits.webkit.org/259916@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3384a016d8f5fb35837403ca3aeb117c7b9cee7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15220 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39002 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115351 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16662 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6405 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115037 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12679 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95648 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40212 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/109634 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94552 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81897 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8466 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28643 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5198 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14582 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48189 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6851 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10502 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -215,10 +215,9 @@ INCLUDED_MIGRATED_HEADERS_YES_cocoatouch = WebEventRegion.h;
-WK_MIGRATED_HEADER_OUTPUT_DIR_ = $(HEADER_OUTPUT_DIR);
-// On non-Mac platforms, all migrated headers go to PrivateHeaders.
-WK_MIGRATED_HEADER_OUTPUT_DIR_cocoatouch = $(HEADER_OUTPUT_DIR:dir)PrivateHeaders;
+WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR = $(WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR_$(WK_COCOA_TOUCH));
```
`WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR ` is only directly used in `MigratedHeaders-output.xcfilelist`. Looking at the PR, I understand that `migrate-headers.py` uses it via the output list, but it's not our usual pattern to directly use xcfilelist content. It would help to have a comment explaining why this section is not dead code.

aproskuryakov commented on Configurations/WebKitProjectPaths.xcconfig:
```diff
@@ -112,6 +112,16 @@ WEBCORE_LIBRARY_DIR[sdk=macosx*] = $(WEBCORE_LIBRARY_DIR_USE_OVERRIDE_FRAMEWORKS
+WEBKITLEGACY_PRIVATE_HEADERS_DIR_Production_COCOA_TOUCH_NO = $(WEBKITLEGACY_PRIVATE_HEADERS_DIR_Production_$(WK_DEFAULT_PLATFORM_NAME)_USE_OVERRIDE_FRAMEWORKS_DIR_$(WK_USE_OVERRIDE_FRAMEWORKS_DIR));
+WEBKITLEGACY_PRIVATE_HEADERS_DIR_Production_macosx_USE_OVERRIDE_FRAMEWORKS_DIR_NO = $(UMBRELLA_FRAMEWORKS_DIR)/WebKitLegacy.framework/$(WK_FRAMEWORK_VERSION_PREFIX)PrivateHeaders;
+WEBKITLEGACY_PRIVATE_HEADERS_DIR_Production_macosx_USE_OVERRIDE_FRAMEWORKS_DIR_YES = $(WK_OVERRIDE_FRAMEWORKS_DIR)/WebKitLegacy.framework/$(WK_FRAMEWORK_VERSION_PREFIX)PrivateHeaders;
+WEBKITLEGACY_PRIVATE_HEADERS_engineering = $(BUILT_PRODUCTS_DIR)/WebKitLegacy.framework/$(WK_FRAMEWORK_VERSION_PREFIX)PrivateHeaders;
```
There is no mistake here, but should this be renamed to `WEBKITLEGACY_PRIVATE_HEADERS_DIR_engineering` for consistency?

emw-apple replied:
The lines I based this off of do the same thing, but I agree that it's oddly named. Will update them all.

aproskuryakov commented on Source/WebKit/MigratedHeaders-input.xcfilelist:
```diff
@@ -0,0 +1,281 @@
+# copied via the "Migrate WebCore and WebKitLegacy Headers" script phase
+# instead of the native Headers phase. 
+
+# The script matches each line of input to the corresponding line in
+# MigratedHeaders-output.xcfilelist, so the order of file names must match.
```
Can the script verify that file names do match?

emw-apple replied:
Good idea!

aproskuryakov reviewed (approved):
This is super elegant!

Just to double check that I understand correctly, `INCLUDED_SOURCE_FILE_NAMES` end up getting passed to `tapi` even though they are not in `xcodeproj`, correct? But how does it know which ones are private and which ones are public headers?

I think that it's nice that the project no longer lists reexported headers, as there is no need to look at them there, but I guess we'll learn if there are downsides.

emw-apple committed:
[Xcode] Move header migration logic to a script phase
https://bugs.webkit.org/show_bug.cgi?id=251635

Reviewed by Alexey Proskuryakov.

WebKit and WebKitLegacy "migrate" headers from other frameworks which
they reexport symbols from. On Mac, some of these migrated headers are
public (namely the WebKitLegacy API on that platform), but on all
embedded platforms the migrated headers are private. Xcode doesn't
natively support headers that have different visibility per-platform.

Previously, we cheated by using a build setting to "redirect" migrated
headers into the PrivateHeaders/ directory on embedded platforms.
Unfortunately, this breaks tapi's SRCROOT mode, so we need to implement
a different workaround.

Replace the migrated header references in both projects with a script
phase that calls a new script, migrate-headers.py. This script reads
from a filelist to process headers and write them to the appropriate
headers folder. Headers that are written to
$(WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR) are the ones with per-platform
visibility.

* Configurations/WebKitProjectPaths.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/MigratedHeaders-input.xcfilelist: Added.
* Source/WebKit/MigratedHeaders-output.xcfilelist: Added.
* Source/WebKit/Scripts/migrate-headers.py: Added.
* Source/WebKit/Scripts/postprocess-header-rule: Now, migrated headers
  that this script sees come from
  $(BUILT_PRODUCTS_DIR)/WebKitLegacy.framework, instead of
  WebKitLegacy's SRCROOT. This means they have already been
  postprocessed by WebKitLegacy. Remove the logic that is duplicated by
  WebKitLegacy's postprocessing script.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259916@main
